### PR TITLE
[DCP] Add CheckpointableTensor protocol

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #189945
* __->__ #189492

The tensor does not need to be wrapped or subclassed. If it provides 
* `global_shape`
* `global_offsets`
* `local_offsets`
* `local_sizes`

DCP can:
  - write checkpoint metadata with the correct global tensor shape
  - create shard write items at the right global offsets
  - map load requests back to the corresponding local tensor slice

focus on default dcp save/load in this PR and intentionally scope out 2 things
* HF safetensor: see 2nd PR in ghstack
* DynamicMetaLoadPlanner / BroadcastingTorchSaveReader: no plan to support since I cannot find callsites

### CODE REVIEW

File: torch/distributed/checkpoint/protocol.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            tensor; needed when one tensor stores multiple shards or includes
+            padding.
+        local_sizes: Shape of each local shard; needed to build checkpoint
+            chunks and slice the local tensor during load.
+    """
+
+    global_shape: tuple[int, ...]
+    global_offsets: tuple[tuple[int, ...], ...]
+    local_offsets: tuple[tuple[int, ...], ...]
+    local_sizes: tuple[tuple[int, ...], ...]
```
Comment (tianyu-l) [inline torch/distributed/checkpoint/protocol.py:38]: What's the consideration of making these attributes vs. methods (e.g. `get_global_shape`)?

File: torch/distributed/checkpoint/protocol.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import torch
+
+from .metadata import ChunkStorageMetadata, MetadataIndex
+
+
+__all__ = ["CheckpointableTensor"]
+
+
+@runtime_checkable
+class CheckpointableTensor(Protocol):
```
Comment (tianyu-l) [inline torch/distributed/checkpoint/protocol.py:14]: Does it / Can we support duck-typing?

File: torch/distributed/checkpoint/protocol.py (hunk 1 above)
Comment (weifengpy) [reply torch/distributed/checkpoint/protocol.py:38]: putting DCP aside, I want flex_shard and spmd_types to follow the same convention: torch.Tensor + attributes

adding DCP into the picture, we already use attributes during forward/backward, I was thinking of keeping using attributes. 

say `get_global_shape ` style, I suspect what it does is `return self._global_shape`

File: torch/distributed/checkpoint/protocol.py (hunk 2 above)
Comment (weifengpy) [reply torch/distributed/checkpoint/protocol.py:14]: duck-typing works and is necessary since we are talking about torch.Tensor attributes.

I happend to verify duck-typing in unit test
```
tensor.global_shape = (16,)
tensor.global_offsets = ((0,), (12,))
tensor.local_offsets = ((0,), (4,))
tensor.local_sizes = ((4,), (4,))
self.assertIsInstance(tensor, CheckpointableTensor)
```

File: torch/distributed/checkpoint/protocol.py (hunk 1 above)
Comment (tianyu-l) [reply torch/distributed/checkpoint/protocol.py:38]: > I want flex_shard and spmd_types to follow the same convention: torch.Tensor + attributes

Makes sense.

Btw does spmd_types have such a protocol? cc @pianpwk

File: torch/distributed/checkpoint/protocol.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    these fields is enough for DCP to map one or more slices of that tensor
+    into a logical global tensor.
+
+    Attributes:
+        global_shape: Full logical tensor shape to write into checkpoint
+            metadata; needed because ``tensor.size()`` is only the local buffer
+            size.
+        global_offsets: Global start coordinate for each local shard; needed to
+            name checkpoint chunks and match load requests by global offset.
+        local_offsets: Start coordinate for each local shard inside the local
```
Comment (wconstab) [inline torch/distributed/checkpoint/protocol.py:28]: local_offsets is a new concept inspired by FlexShard, right?  I think you could improve the doc for how it is allowed to be used:

Is it required that local_* refer to shards all part of this global tensor?  What if the local tensor is a flat buffer containing shards of multiple global tensors?  Is it OK that local_* fields refer to non-complete parts of the local tensor?

There are a couple related ways to do things-
- (a) slice views out of a tensor, then checkpoitn each view: could get rid of local_offsets/local_sizes in this use case
- (b) checkpoint a flat buffer that contains multiple shards of a tensor: needs local_* fields here
- (c) checkpoint a flat buffer that contains multiple shards of **multiuple** tensors- seems unsupported by this version of local_ and global_ metadata


Do you agree with the characterization of a,b,c? and if so, is the argument that (b) matters but (c) does not?

File: torch/distributed/checkpoint/protocol.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            chunks and slice the local tensor during load.
+    """
+
+    global_shape: tuple[int, ...]
+    global_offsets: tuple[tuple[int, ...], ...]
+    local_offsets: tuple[tuple[int, ...], ...]
+    local_sizes: tuple[tuple[int, ...], ...]
+
+
+def _is_checkpointable_tensor(obj: object) -> bool:
```
Comment (Skylion007) [inline torch/distributed/checkpoint/protocol.py]: This is a TypeGuard[CheckpointableTensor] return. Let's Python typing know that if the Boolean is true it satisfies the protocol and can narrow accordingly.
```suggestion
def _is_checkpointable_tensor(obj: object) -> TypeGuard[CheckpointableTensor]:

File: torch/distributed/checkpoint/protocol.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    return isinstance(obj, torch.Tensor) and isinstance(obj, CheckpointableTensor)
+
+
+def _get_checkpointable_tensor_chunks(
+    tensor: torch.Tensor,
+) -> list[ChunkStorageMetadata]:
+    _validate_checkpointable_tensor_metadata(tensor)
+    checkpointable_tensor = cast(CheckpointableTensor, tensor)
```
Comment (Skylion007) [inline torch/distributed/checkpoint/protocol.py]: After the fix above can just do an assert use function above
```suggestion
    checkpointable_tensor = tensor
```

File: torch/distributed/checkpoint/protocol.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        raise ValueError("global_offsets and local_sizes must have the same length")
+
+    global_shape = checkpointable_tensor.global_shape
+    tensor_shape = tuple(tensor.size())
+    for idx, (global_offset, local_offset, local_size) in enumerate(
+        zip(
+            checkpointable_tensor.global_offsets,
+            checkpointable_tensor.local_offsets,
+            checkpointable_tensor.local_sizes,
+        )
```
Comment (Skylion007) [inline torch/distributed/checkpoint/protocol.py:134]: Strict=True?

File: torch/distributed/checkpoint/protocol.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        for dim, (offset, size, global_dim) in enumerate(
+            zip(global_offset, local_size, global_shape)
+        ):
+            if offset < 0 or size < 0 or offset + size > global_dim:
+                raise ValueError(
+                    f"global shard {idx} dimension {dim} is outside global_shape"
+                )
+
+        for dim, (offset, size, local_dim) in enumerate(
+            zip(local_offset, local_size, tensor_shape)
```
Comment (Skylion007) [inline torch/distributed/checkpoint/protocol.py]: should also be strict=True kwarg?

File: torch/distributed/checkpoint/default_planner.py [hunk 8]
```diff
@@ -500,11 +504,16 @@ def create_default_local_load_plan(
         if (
             isinstance(md, TensorStorageMetadata)
             and getattr(obj, "size", None) is not None
-            and md.size != obj.size()
         ):
-            raise ValueError(
-                f"Size mismatch between saved {md.size} and current: {obj.size()} for {fqn}",
+            obj_size = (
+                torch.Size(cast(CheckpointableTensor, obj).global_shape)
```
Comment (Skylion007) [inline torch/distributed/checkpoint/default_planner.py]: The cast becomes unnecessary after the TypeGuard fix.

File: torch/distributed/checkpoint/protocol.py (hunk 3 above)
Comment (weifengpy) [reply torch/distributed/checkpoint/protocol.py:28]: > Is it required that local_* refer to shards all part of this global tensor? 

Yes. For `state_dict = {fqn: tensor}`, `fqn` names a single global tensor (parameter). tensor itself can be viewed from of a flat buffer. But the `CheckpointableTensor` metadata for that tensor (including `local_offsets`, `local_sizes`) only describes regions inside one global tensor (parameter)


(will answer the rest of the question when I am near to laptop)

Comment (weifengpy) [reply torch/distributed/checkpoint/protocol.py:28]: > Is it OK that local_* fields refer to non-complete parts of the local tensor?

`local_*` can refer to non-complete parts. The only requriement is it's describing a single logical global tensor named by this FQN. flat buffer containing multiple parameters isn't visible to DCP. This is same to above reply

> (a) slice views out of a tensor, then checkpoitn each view: could get rid of local_offsets/local_sizes in this use case

I think you are referring to viewing out from a flat buffer - that's done inside `model.state_dict()`. This `local_*` is mostly referring to chunks within a parameter 

if you are referring to viewing out from a single parameter, local_offsets/local_sizes can be replaced by `local_tensors: List[torch.Tensor]`


> (b) checkpoint a flat buffer that contains multiple shards of a tensor: needs local_* fields here

each parameter shard is always contiguous in the flat buffer currently. I would say (b) is intentionally not supported 

> (c) checkpoint a flat buffer that contains multiple shards of multiuple tensors- seems unsupported by this version of local_ and global_ metadata

characterization is correct - this is (intentionally) not unsupported

File: torch/distributed/checkpoint/protocol.py (hunk 4 above)
Comment (weifengpy) [reply torch/distributed/checkpoint/protocol.py]: I updated the PR following your suggestions (including other suggestions). 
thanks for holding the bar - the code looks better

**Review by meetv18 — APPROVED** [review]
